### PR TITLE
Woche5

### PR DESCRIPTION
**Pull request by Koch-Dustin**
Changes:
   1. Add Inheritance
   2. Add Speller

PR: #7

### CODE REVIEW

File: Woche5/speller/speller.c [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        printf("Could not load %s.\n", dictionary);
+
+        return 1;
+    }
+
+    time_load = calculate(&before, &after);
+
+    char *text = (argc == 3) ? argv[2] : argv[1];
+    FILE *file = fopen(text, "r");
+    if (file == NULL) {
```
Comment (FrederikFricke) [inline Woche5/speller/speller.c:43]:
```suggestion
    char *text = (argc == 3) ? argv[2] : argv[1];
    FILE *file = fopen(text, "r");
    
    if (file == NULL) {
```

**Review by FrederikFricke — CHANGES_REQUESTED** [review]

File: Woche5/inheritance/inheritance.c [hunk 2]
```diff
@@ -0,0 +1,97 @@
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <time.h>
+#include <math.h>
+
+typedef struct person
+{
+    struct person *parents[2];
+    char alleles[2];
```
Comment (FrederikFricke) [inline Woche5/inheritance/inheritance.c:10]:
```suggestion
typedef struct person {
    struct person *parents[2];
    char alleles[2];
```

File: Woche5/speller/dictionary.c [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+            n->next = NULL;
+        }
+
+        table[index] = n;
+
+        total_words++;
+
+    }
+
+    fclose(file);
```
Comment (FrederikFricke) [inline Woche5/speller/dictionary.c:76]:
```suggestion

        table[index] = n;
        total_words++;
    }

    fclose(file);
```

File: Woche5/speller/dictionary.c [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+bool unload(void) {
+    for (int i = 0; i < N; i++) {
+        if(table[i]!= NULL) {
+            destroy(table[i]);
+        }
+    }
+    return true;
+}
```
Comment (FrederikFricke) [inline Woche5/speller/dictionary.c:101]:
```suggestion
    }
    
    return true;
}
```

**Review by FrederikFricke — CHANGES_REQUESTED** [review]

**Review by FrederikFricke — APPROVED** [review]

File: Woche5/speller/dictionary.c [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+    fclose(file);
+
+    return true;
+}
+
+unsigned int size(void) {
+
+    return total_words;
+}
```
Comment (stefanluth) [inline Woche5/speller/dictionary.c:82]:
```suggestion
unsigned int size(void) {
    return total_words;
}
```

**Review by stefanluth — CHANGES_REQUESTED** [review]

File: Woche5/speller/speller.c [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    char c;
+    while (fread(&c, sizeof(char), 1, file)) {
+
+        if (isalpha(c) || (c == '\'' && index > 0)) {
+
+            word[index] = c;
+            index++;
+
+            if (index > LENGTH) {
+
```
Comment (stefanluth) [inline Woche5/speller/speller.c:65]:
```suggestion
    while (fread(&c, sizeof(char), 1, file)) {
        if (isalpha(c) || (c == '\'' && index > 0)) {
            word[index] = c;
            index++;

            if (index > LENGTH) {
```

File: Woche5/speller/speller.c [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+        else if (isdigit(c)) {
+            while (fread(&c, sizeof(char), 1, file) && isalnum(c));
+
+            index = 0;
+        } else if (index > 0) {
+
+            word[index] = '\0';
+
+            words++;
```
Comment (stefanluth) [inline Woche5/speller/speller.c:80]:
```suggestion
            word[index] = '\0';
            words++;
```

File: Woche5/speller/speller.c [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+#include "dictionary.h"
+
+#undef calculate
+#undef getrusage
+
+#define DICTIONARY "dictionaries/large"
+
+double calculate(const struct rusage *b, const struct rusage *a);
+
+int main(int argc, char *argv[]) {
```
Comment (stefanluth) [inline Woche5/speller/speller.c:15]: The main is huge, split it into more functions.

**Review by stefanluth — CHANGES_REQUESTED** [review]

File: Woche5/inheritance/inheritance.c [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+char random_allele() {
+    int randomNumber = rand() % 3;
+
+    if (randomNumber == 0) {
+        return 'A';
+    } else if (randomNumber == 1) {
+        return 'B';
+    } else {
+        return 'O';
+    }
```
Comment (DevBastiS) [inline Woche5/inheritance/inheritance.c:95]: Is it possible to solve this with a switch case?

**Review by DevBastiS — CHANGES_REQUESTED** [review]

**Comment by Koch-Dustin** [comment]
> Screenshot 2022-09-08 at 14 13 34

Should be done

File: Woche5/inheritance/inheritance.c [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+void free_family(person *personPointer) {
+
+    if (personPointer == NULL) return;
+
+    free_family(personPointer -> parents[0]);
+    free_family(personPointer -> parents[1]);
+
+    free(personPointer);
+}
```
Comment (stefanluth) [inline Woche5/inheritance/inheritance.c:64]: Missing returns. Other occurrences as well.

**Review by stefanluth — CHANGES_REQUESTED** [review]

File: Woche5/inheritance/inheritance.c [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+char random_allele() {
+    int randomNumber = rand() % 3;
+
+    switch(randomNumber) {
+	    case 0: printf("A\n"); break;
+	    case 1: printf("B\n"); break;
+	    default: printf("O\n"); break;
+    }
+    return;
+}
```
Comment (stefanluth) [inline Woche5/inheritance/inheritance.c:101]: This function is supposed to return a `char` but it returns `void`.

File: Woche5/inheritance/inheritance.c [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+    srand(time(0));
+
+    person *personPointer = create_family(GENERATIONS);
+
+    print_family(personPointer, 0);
+
+    free_family(personPointer);
+
+    return;
+}
```
Comment (stefanluth) [inline Woche5/inheritance/inheritance.c:32]: This function is supposed to return an `int` but it returns `void`.

File: Woche5/speller/dictionary.c [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+
+bool unload(void) {
+    for (int i = 0; i < N; i++) {
+        if (table[i]!= NULL) {
+            destroy(table[i]);
+        }
+    }
+    
+    return true;
+}
```
Comment (stefanluth) [inline Woche5/speller/dictionary.c:100]: This function always returns `true`.

File: Woche5/speller/speller.c [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+    
+    if (file == NULL) {
+        printf("Could not open %s.\n", text);
+        unload();
+
+        return 1;
+    }
+
+    printf("\nMISSPELLED WORDS\n\n");
+}
```
Comment (stefanluth) [inline Woche5/speller/speller.c:104]: This function is supposed to return `void` but returns `int`. Also a missing return statement at the end.

**Review by stefanluth — CHANGES_REQUESTED** [review]
Please check missing returns and returning the correct types. There are more than I commented.

**Review by DevBastiS — APPROVED** [review]
Wait for @stefanluth's approval.

**Comment by stefanluth** [comment]
Or not :D

**Pull request merged**